src/mce_plugin.c

```diff
@@ -0,0 +1,197 @@
+    MANAGER_EVENT_COUNT
+};
+
+typedef NfcPluginClass McePluginClass;
+typedef struct mce_plugin {
+    NfcPlugin parent;
+    NfcManager* manager;
+    MceDisplay* display;
+    gulong manager_event_id[MANAGER_EVENT_COUNT];
+    gulong display_event_id[DISPLAY_EVENT_COUNT];
+    gboolean always_on;
+} McePlugin;
+
+G_DEFINE_TYPE(McePlugin, mce_plugin, NFC_TYPE_PLUGIN)
+#define THIS_TYPE (mce_plugin_get_type())
+#define THIS(obj) (G_TYPE_CHECK_INSTANCE_CAST((obj), THIS_TYPE, McePlugin))
+
+#define mce_plugin_display_on(display) \
+    ((display) && (display)->valid && (display)->state != MCE_DISPLAY_STATE_OFF)
+
+/* These need to be synchronized with the settings plugin */
+#define SETTINGS_STORAGE_PATH   "/var/lib/nfcd/settings"
+#define SETTINGS_GROUP          "Settings"
+#define SETTINGS_KEY_ALWAYS_ON  "AlwaysOn"
+
+static
+void
+mce_plugin_update_power(
+    McePlugin* self)
+{
+    nfc_manager_request_power(self->manager, self->manager->enabled &&
+        (self->always_on || mce_plugin_display_on(self->display)));
+}
+
+static
+void
+mce_plugin_manager_state_handler(
+    NfcManager* manager,
+    void* plugin)
+{
+    mce_plugin_update_power(THIS(plugin));
+}
+
+static
+void
+mce_plugin_display_state_handler(
+    MceDisplay* display,
+    gpointer plugin)
+{
+    mce_plugin_update_power(THIS(plugin));
+}
+
+static
+gboolean
+mce_plugin_start(
+    NfcPlugin* plugin,
+    NfcManager* manager)
+{
+    McePlugin* self = THIS(plugin);
+
+    GVERBOSE("Starting");
+    GASSERT(!self->manager);
+    self->manager = nfc_manager_ref(manager);
+    self->manager_event_id[MANAGER_ENABLED] =
+        nfc_manager_add_enabled_changed_handler(manager,
+            mce_plugin_manager_state_handler, self);
+
+    /* No need to track the display state if we are always on */
+    if (!self->always_on) {
+        self->display = mce_display_new();
+        self->display_event_id[DISPLAY_VALID] =
+            mce_display_add_valid_changed_handler(self->display,
+                mce_plugin_display_state_handler, self);
+        self->display_event_id[DISPLAY_STATE] =
+            mce_display_add_state_changed_handler(self->display,
+                mce_plugin_display_state_handler, self);
+    }
+    mce_plugin_update_power(self);
+    return TRUE;
+}
+
+static
+void
+mce_plugin_stop(
+    NfcPlugin* plugin)
+{
+    McePlugin* self = THIS(plugin);
+
+    GVERBOSE("Stopping");
+    if (self->display) {
+        mce_display_remove_all_handlers(self->display, self->display_event_id);
+        mce_display_unref(self->display);
+        self->display = NULL;
+    }
+    nfc_manager_remove_all_handlers(self->manager, self->manager_event_id);
+    nfc_manager_unref(self->manager);
+    self->manager = NULL;
+}
+
+static
+void
+mce_plugin_init(
+    McePlugin* self)
+{
+    GKeyFile* config = g_key_file_new();
+
+    if (g_key_file_load_from_file(config, SETTINGS_STORAGE_PATH, 0, NULL)) {
+        self->always_on = g_key_file_get_boolean(config, SETTINGS_GROUP,
```

Comment:
Previously: AFAICT setting value was accessed via settings plugin which had some logical preparations for making runtime changes and updating config file vs. now: the value is read directly from config when mce plugin is loaded and no changes are anticipated. Which is just fine as long as that is the intent but I'm a bit confused/unsure, so maybe you could clarify this a bit?

Here it seems that the idea is that the config file is accessed directly when mce plugin is initialized and plugin locks on to whatever the value at init time. However

Reply:
It's basically nothing more than a debugging trick, to keep nfc on even when display turns off. This option is not exposed to the user, not modifiable via D-Bus or anything. You can only put it there by hand-editing the settings file. Once it's there, it's preserved by the built-in settings plugin.